File: load-test/dify_stress_test.js
// Dify 平台 5000 人并发压测 (k6)
// 模拟高职院校真实教学场景：学生上课同时使用 AI 助手
//
// Usage:
//   docker run --rm -v /tmp:/scripts -w /scripts grafana/k6 run \
//     -e BASE_URL=https://10.167.2.175:31825 dify_stress_test.js
//
// 阶梯式加压: 100 -> 500 -> 1000 -> 2000 -> 5000 VU

import http from 'k6/http';
import { check, sleep } from 'k6';
import { Counter, Rate, Trend } from 'k6/metrics';
import encoding from 'k6/encoding';

const BASE_URL = __ENV.BASE_URL || 'https://10.167.2.175:31825';
const EMAIL = __ENV.EMAIL || 'myuwei@126.com';
const PASSWORD = __ENV.PASSWORD || 'Difyai123456';

// 自定义指标
const chatSuccessRate = new Rate('chat_success_rate');
const chatDuration = new Trend('chat_duration_ms', true);
const platformDuration = new Trend('platform_duration_ms', true);
const streamingChunks = new Counter('streaming_chunks');

// 高职院校真实教学提问池
const TEACHING_QUESTIONS = [
  '什么是Python中的变量？',
  '请解释一下二叉树的遍历方式',
  'Java和Python哪个更适合初学者？',
  '什么是数据库的事务？',
  'TCP和UDP的区别是什么？',
  '请用C语言写一个冒泡排序',
  '什么是机器学习中的过拟合？',
  '如何理解面向对象的封装特性？',
  'SQL中的GROUP BY怎么用？',
  'HTTP状态码404是什么意思？',
  '什么是云计算的IaaS、PaaS、SaaS？',
  'JavaScript的let和var有什么区别？',
  '网络子网掩码怎么计算？',
  '什么是软件工程中的敏捷开发？',
  '请解释CSS盒模型',
  '数据结构中栈和队列的区别？',
  '什么是RESTful API？',
  'Linux常用命令有哪些？',
  'Python的列表和元组有什么区别？',
  '什么是大数据的4V特征？',
];

export const options = {
  scenarios: {
    // 场景1: 平台基础能力测试 (不依赖 LLM 推理) - 测试 HPA/ingress/DB 层
    platform: {
      executor: 'ramping-vus',
      startVUs: 0,
      stages: [
        { duration: '1m', target: 500 },
        { duration: '2m', target: 2000 },
        { duration: '3m', target: 5000 },
        { duration: '3m', target: 5000 },
        { duration: '1m', target: 0 },
      ],
      gracefulRampDown: '30s',
      exec: 'platformTest',
    },
    // 场景2: LLM 聊天测试 (限制并发避免 LiteLLM 超时)
    llmChat: {
      executor: 'ramping-vus',
      startVUs: 0,
      startTime: '30s',
      stages: [
        { duration: '1m', target: 50 },
        { duration: '2m', target: 100 },
        { duration: '3m', target: 200 },
        { duration: '2m', target: 100 },
        { duration: '1m', target: 0 },
      ],
      gracefulRampDown: '30s',
      exec: 'llmChatTest',
    },
  },
  thresholds: {
    'http_req_failed': ['rate<0.15'],
    'platform_duration_ms': ['p(95)<3000'],
    'chat_success_rate': ['rate>0.70'],
  },
  insecureSkipTLSVerify: true,
};

const vuState = {};

export function setup() {
  // 预登录获取应用信息
  const headers = { 'Host': 'console.dify-plus.local', 'Content-Type': 'application/json' };
  const passB64 = encodeBase64(PASSWORD);
  const loginResp = http.post(`${BASE_URL}/console/api/login`,
    JSON.stringify({ email: EMAIL, password: passB64, language: 'zh-Hans', remember_me: true }),
    { headers, insecureSkipTLSVerify: true });

  let appToken = null;
  let inputs = {};
  if (loginResp.status === 200) {
    const csrf = loginResp.cookies['csrf_token'] ? loginResp.cookies['csrf_token'][0].value : '';
    const authHeaders = { 'Host': 'console.dify-plus.local', 'X-CSRF-Token': csrf };
    http.get(`${BASE_URL}/console/api/apps?page=1&page_size=50`, { headers: authHeaders });

    // 获取 API key
    const keysResp = http.get(`${BASE_URL}/console/api/apps`, { headers: authHeaders });
    if (keysResp.status === 200) {
      const apps = keysResp.json('data') || [];
      const chatApp = apps.find(a => a.mode === 'chat');
      if (chatApp) {
        const keyResp = http.get(`${BASE_URL}/console/api/apps/${chatApp.id}/api-keys`, { headers: authHeaders });
        if (keyResp.status === 200) {
          const keys = keyResp.json('data') || [];
          if (keys.length > 0) {
            appToken = keys[0].token;
          }
        }
        // 获取参数
        if (appToken) {
          const paramResp = http.get(`${BASE_URL}/v1/parameters`, {
            headers: { 'Host': 'api.dify-plus.local', 'Authorization': `Bearer ${appToken}` },
            insecureSkipTLSVerify: true,
          });
          if (paramResp.status === 200) {
            const form = paramResp.json('user_input_form') || [];
            for (const item of form) {
              for (const [type, cfg] of Object.entries(item)) {
                if (cfg.required && cfg.variable) {
                  if (type === 'select') {
                    inputs[cfg.variable] = (cfg.options && cfg.options[0]) || 'True';
                  } else {
                    inputs[cfg.variable] = '老师讲课很认真，内容丰富，受益匪浅';
                  }
                }
              }
            }
          }
        }
      }
    }
  }

  return { appToken, inputs, csrf: loginResp.cookies['csrf_token'] ? loginResp.cookies['csrf_token'][0].value : '', sessionCookie: `access_token=${loginResp.cookies['access_token'] ? loginResp.cookies['access_token'][0].value : ''}` };
}

function encodeBase64(str) {
  // k6 的 encoding.b64encode
  return encoding.b64encode(str);
}

// 平台基础能力测试：5000并发，不依赖LLM推理（测试 ingress/api/db/redis 层）
export function platformTest(data) {
  const studentId = `student-${__VU}-${__ITER}`;
  const csrfHeaders = {
    'Host': 'console.dify-plus.local',
    'X-CSRF-Token': data.csrf || '',
    'Cookie': data.sessionCookie || '',
  };

  // 1. 应用列表 (测试 DB 查询 + API 响应)
  const appsResp = http.get(`${BASE_URL}/console/api/apps?page=1&page_size=20`, {
    headers: csrfHeaders,
    tags: { name: 'apps-list' },
  });
  check(appsResp, { 'apps 200': (r) => r.status === 200 });

  // 2. 知识库列表 (测试 DB 查询)
  const dsResp = http.get(`${BASE_URL}/console/api/datasets?page=1&page_size=20`, {
    headers: csrfHeaders,
    tags: { name: 'datasets-list' },
  });
  check(dsResp, { 'datasets 200': (r) => r.status === 200 });

  // 3. 用户资料 (测试 Redis 缓存)
  const profileResp = http.get(`${BASE_URL}/console/api/account/profile`, {
    headers: csrfHeaders,
    tags: { name: 'profile' },
  });
  check(profileResp, { 'profile 200': (r) => r.status === 200 });
  platformDuration.add(profileResp.timings.waiting);

  // 4. 如果有 app token，测试对话列表 (不发起 LLM 推理)
  if (data.appToken) {
    const apiHeaders = {
      'Host': 'api.dify-plus.local',
      'Authorization': `Bearer ${data.appToken}`,
    };
    const convResp = http.get(`${BASE_URL}/v1/conversations?user=${studentId}&limit=5`, {
      headers: apiHeaders,
      tags: { name: 'conversations-list' },
    });
    check(convResp, { 'conversations 200': (r) => r.status === 200 });
    platformDuration.add(convResp.timings.waiting);
  }

  sleep(Math.random() * 2 + 0.5);
}

// LLM 聊天测试：限制并发到200，测试 LiteLLM 网关推理能力
export function llmChatTest(data) {
  if (!data.appToken) {
    sleep(5);
    return;
  }

  const studentId = `llm-student-${__VU}`;
  const query = TEACHING_QUESTIONS[Math.floor(Math.random() * TEACHING_QUESTIONS.length)];
  const headers = {
    'Host': 'api.dify-plus.local',
    'Authorization': `Bearer ${data.appToken}`,
    'Content-Type': 'application/json',
  };

  const payload = {
    inputs: data.inputs,
    query: query,
    response_mode: 'blocking',
    user: studentId,
  };

  const startTime = new Date().getTime();
  const resp = http.post(`${BASE_URL}/v1/chat-messages`, JSON.stringify(payload), {
    headers,
    tags: { name: 'chat-blocking' },
    timeout: '300s',
  });
  const duration = new Date().getTime() - startTime;

  const success = check(resp, {
    'chat 200': (r) => r.status === 200,
  });

  chatSuccessRate.add(success);
  chatDuration.add(duration);

  sleep(Math.random() * 5 + 3);
}

export default function (data) {
  // default 不执行（由 platformTest 和 llmChatTest 分场景执行）
}

export function handleSummary(data) {
  const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
  return {
    [`/tmp/dify_stress_summary_${timestamp}.json`]: JSON.stringify(data.metrics, null, 2),
    stdout: textSummary(data),
  };
}

function textSummary(data) {
  const m = data.metrics;
  let out = '\n' + '='.repeat(70) + '\n';
  out += 'Dify 5000人并发压测结果\n' + '='.repeat(70) + '\n';

  if (m.http_req_duration) {
    out += `\nHTTP 响应时间:\n`;
    out += `  P50: ${m.http_req_duration.values['p(50)']?.toFixed(0) || 'N/A'} ms\n`;
    out += `  P90: ${m.http_req_duration.values['p(90)']?.toFixed(0) || 'N/A'} ms\n`;
    out += `  P95: ${m.http_req_duration.values['p(95)']?.toFixed(0) || 'N/A'} ms\n`;
  }
  if (m.http_reqs) {
    out += `\n总请求数: ${m.http_reqs.values.count}\n`;
    out += `吞吐量: ${(m.http_reqs.values.rate || 0).toFixed(1)} req/s\n`;
  }
  if (m.http_req_failed) {
    out += `失败率: ${((m.http_req_failed.values.rate || 0) * 100).toFixed(2)}%\n`;
  }
  if (m.chat_success_rate) {
    out += `聊天成功率: ${((m.chat_success_rate.values.rate || 0) * 100).toFixed(2)}%\n`;
  }
  if (m.vus_max) {
    out += `最大并发用户: ${m.vus_max.values.value}\n`;
  }
  out += '\n' + '='.repeat(70) + '\n';
  return out;
}

File: load-test/dify_stress_test.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Dify 平台 5000 人并发压测脚本 (Locust)
=========================================
模拟真实教学场景: 学生上课同时使用 AI 助手
- Dify 平台用户: 登录 -> 应用列表 -> 发送聊天 -> 流式聊天 -> 对话历史
- LiteLLM 网关用户: health -> models -> chat completions -> embeddings

Usage:
  locust -f dify_stress_test.py --host=https://10.167.2.175:31825 --headless \
         -u 5000 -r 100 --run-time 10m --html dify_stress_report.html
"""
import json
import base64
import random
import string
from locust import HttpUser, task, between, tag, events

# 配置
EMAIL = "myuwei@126.com"
PASSWORD = "Difyai123456"
LITELLM_HOST = "http://10.167.2.176:30083"
LITELLM_KEY = "sk-ai-platform-master"

# 高职院校真实教学提问池
TEACHING_QUESTIONS = [
    "什么是Python中的变量？",
    "请解释一下二叉树的遍历方式",
    "Java和Python哪个更适合初学者？",
    "什么是数据库的事务？",
    "TCP和UDP的区别是什么？",
    "请用C语言写一个冒泡排序",
    "什么是机器学习中的过拟合？",
    "如何理解面向对象的封装特性？",
    "SQL中的GROUP BY怎么用？",
    "HTTP状态码404是什么意思？",
    "什么是云计算的IaaS、PaaS、SaaS？",
    "JavaScript的let和var有什么区别？",
    "网络子网掩码怎么计算？",
    "什么是软件工程中的敏捷开发？",
    "请解释CSS盒模型",
    "数据结构中栈和队列的区别？",
    "什么是RESTful API？",
    "Linux常用命令有哪些？",
    "什么是区块链技术？",
    "Python的列表和元组有什么区别？",
]


class DifyStudentUser(HttpUser):
    """模拟学生使用 Dify AI 助手 - 5000 并发"""
    weight = 7
    wait_time = between(2, 8)

    def on_start(self):
        """登录获取 session"""
        self.app_id = None
        self.app_token = None
        self.conversation_id = None
        self.student_id = f"student-{random.randint(1, 9999)}"
        self.login()
        self.pick_app()

    def login(self):
        """管理员登录（模拟学生通过已发布应用访问，此处模拟控制台登录）"""
        pass_b64 = base64.b64encode(PASSWORD.encode()).decode()
        with self.client.post(
            "/console/api/login",
            json={"email": EMAIL, "password": pass_b64,
                  "language": "zh-Hans", "remember_me": True},
            name="/console/api/login",
            catch_response=True,
        ) as resp:
            if resp.status_code == 200:
                try:
                    csrf = resp.cookies.get("csrf_token", "")
                    self.client.headers.update({"X-CSRF-Token": csrf})
                    resp.success()
                except Exception as e:
                    resp.failure(f"parse error: {e}")
            else:
                resp.failure(f"login failed: {resp.status_code}")

    def pick_app(self):
        """随机选择一个聊天应用"""
        with self.client.get(
            "/console/api/apps?page=1&page_size=50",
            name="/console/api/apps",
            catch_response=True,
        ) as resp:
            if resp.status_code == 200:
                apps = resp.json().get("data", [])
                chat_apps = [a for a in apps if a.get("mode") == "chat"]
                if chat_apps:
                    self.app_id = random.choice(chat_apps).get("id")
                    self.get_api_key()
                resp.success()
            else:
                resp.failure(f"apps failed: {resp.status_code}")

    def get_api_key(self):
        """获取应用 API Key"""
        if not self.app_id:
            return
        with self.client.get(
            f"/console/api/apps/{self.app_id}/api-keys",
            name="/console/api/apps/api-keys",
            catch_response=True,
        ) as resp:
            if resp.status_code == 200:
                keys = resp.json().get("data", [])
                if keys:
                    self.app_token = keys[0].get("token")
                else:
                    r = self.client.post(
                        f"/console/api/apps/{self.app_id}/api-keys",
                        name="/console/api/apps/api-keys-create")
                    self.app_token = r.json().get("token", "")
                resp.success()
            else:
                resp.failure(f"api-keys failed: {resp.status_code}")

    @task(5)
    def send_chat_blocking(self):
        """发送聊天消息（阻塞模式）- 教学提问"""
        if not self.app_token:
            self.pick_app()
            return
        query = random.choice(TEACHING_QUESTIONS)
        headers = {
            "Authorization": f"Bearer {self.app_token}",
            "Host": "api.dify-plus.local",
            "Content-Type": "application/json",
        }
        payload = {
            "inputs": {"Text1": query},
            "query": query,
            "response_mode": "blocking",
            "user": self.student_id,
        }
        if self.conversation_id:
            payload["conversation_id"] = self.conversation_id
        with self.client.post(
            "/v1/chat-messages",
            json=payload,
            headers=headers,
            name="/v1/chat-messages [blocking]",
            catch_response=True,
            timeout=120,
        ) as resp:
            if resp.status_code == 200:
                data = resp.json()
                self.conversation_id = data.get("conversation_id")
                resp.success()
            elif resp.status_code == 400:
                # 模型超时等预期内失败也算部分成功（标记但不计入失败率）
                resp.failure(f"chat 400: {resp.text[:80]}")
            else:
                resp.failure(f"chat failed: {resp.status_code}")

    @task(2)
    def send_chat_streaming(self):
        """流式聊天 - 模拟学生等待打字机效果"""
        if not self.app_token:
            return
        query = random.choice(TEACHING_QUESTIONS)
        headers = {
            "Authorization": f"Bearer {self.app_token}",
            "Host": "api.dify-plus.local",
        }
        payload = {
            "inputs": {"Text1": query},
            "query": query,
            "response_mode": "streaming",
            "user": self.student_id,
        }
        with self.client.post(
            "/v1/chat-messages",
            json=payload,
            headers=headers,
            name="/v1/chat-messages [streaming]",
            catch_response=True,
            timeout=120,
            stream=True,
        ) as resp:
            if resp.status_code == 200:
                chunk_count = 0
                for line in resp.iter_lines():
                    if line and line.startswith(b"data:"):
                        chunk_count += 1
                resp.success()
            else:
                resp.failure(f"stream failed: {resp.status_code}")

    @task(1)
    def view_conversations(self):
        """查看对话历史"""
        if not self.app_token:
            return
        headers = {
            "Authorization": f"Bearer {self.app_token}",
            "Host": "api.dify-plus.local",
        }
        with self.client.get(
            "/v1/conversations",
            params={"user": self.student_id, "limit": 10},
            headers=headers,
            name="/v1/conversations",
            catch_response=True,
        ) as resp:
            if resp.status_code == 200:
                resp.success()
            else:
                resp.failure(f"conversations failed: {resp.status_code}")


class LiteLLMGatewayUser(HttpUser):
    """模拟直接调用 LiteLLM 网关的 AI 服务 - 5000 并发"""
    weight = 3
    wait_time = between(1, 5)

    def on_start(self):
        self.model = random.choice(["qwen2.5:7b", "qwen2.5:14b", "qwen2.5-coder:7b"])

    @task(6)
    def chat_completion(self):
        """LLM 聊天补全"""
        headers = {"Authorization": f"Bearer {LITELLM_KEY}"}
        with self.client.post(
            "/v1/chat/completions",
            json={
                "model": self.model,
                "messages": [{"role": "user", "content": random.choice(TEACHING_QUESTIONS)}],
                "max_tokens": 50,
            },
            headers=headers,
            name="/v1/chat/completions",
            catch_response=True,
            timeout=90,
        ) as resp:
            if resp.status_code == 200:
                resp.success()
            else:
                resp.failure(f"litellm chat: {resp.status_code}")

    @task(2)
    def list_models(self):
        """获取模型列表"""
        headers = {"Authorization": f"Bearer {LITELLM_KEY}"}
        with self.client.get(
            "/v1/models",
            headers=headers,
            name="/v1/models",
            catch_response=True,
        ) as resp:
            if resp.status_code == 200:
                resp.success()
            else:
                resp.failure(f"models: {resp.status_code}")

    @task(1)
    def health_check(self):
        """健康检查"""
        with self.client.get(
            "/health/liveliness",
            name="/health/liveliness",
            catch_response=True,
        ) as resp:
            if resp.status_code == 200:
                resp.success()
            else:
                resp.failure(f"health: {resp.status_code}")
